airbyte-cdk/python/airbyte_cdk/utils/airbyte_secrets_utils.py

```diff
@@ -2,24 +2,43 @@
 # Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 #
 
-import logging
-from typing import TYPE_CHECKING, Any, List, Mapping
+from typing import Any, List, Mapping
 
-from airbyte_cdk.utils.mapping_utils import all_key_pairs_dot_notation, get_value_by_dot_notation
+import dpath.util
 
-if TYPE_CHECKING:
-    from airbyte_cdk.sources import Source
 
+def get_secret_paths(schema: Mapping[str, Any]) -> List[str]:
+    paths = []
 
-def get_secrets(source: "Source", config: Mapping[str, Any], logger: logging.Logger) -> List[Any]:
+    def traverse_schema(schema: Any, path: List[str]):
+        if isinstance(schema, dict):
+            for k, v in schema.items():
+                traverse_schema(v, [*path, k])
+        elif isinstance(schema, list):
+            for i in schema:
+                traverse_schema(i, path)
+        else:
+            if path[-1] == "airbyte_secret" and schema is True:
+                filtered_path = [p for p in path[:-1] if p not in ["properties", "oneOf"]]
+                paths.append(filtered_path)
+
+    traverse_schema(schema, [])
+    return paths
+
+
+def get_secrets(connection_specification: Mapping[str, Any], config: Mapping[str, Any]) -> List[Any]:
     """
-    Get a list of secrets from the source config based on the source specification
+    Get a list of secret values from the source config based on the source specification
+    :type connection_specification: the connection_specification field of an AirbyteSpecification i.e the JSONSchema definition
     """
-    flattened_key_values = all_key_pairs_dot_notation(source.spec(logger).connectionSpecification.get("properties", {}))
-    secret_key_names = [
-        ".".join(key.split(".")[:1]) for key, value in flattened_key_values.items() if value and key.endswith("airbyte_secret")
-    ]
-    return [str(get_value_by_dot_notation(config, key)) for key in secret_key_names if config.get(key)]
+    secret_paths = get_secret_paths(connection_specification.get("properties", {}))
+    result = []
+    for path in secret_paths:
+        try:
+            result.append(dpath.util.get(config, path))
+        except KeyError:
+            pass
```

Comment:
why would this fail in a way that we can ignore?

Reply:
`dpath` is a library for getting/setting deeply nested fields in dicts. In the case of `oneOf`s, we check for all the secret so we don't expect to find some fields. When we don't find them, we expect a `KeyError`. I'll add a comment clarifying this.

Reply:
added this comment

```
"""
        schema_item can be any property or value in the originally input jsonschema, depending on how far down the recursion stack we go
        path is the path to that schema item in the original input
        for example if we have the input {'password': {'type': 'string', 'airbyte_secret': True}} then the arguments will evolve
        as follows:
        schema_item=<whole_object>, path=[]
        schema_item={'type': 'string', 'airbyte_secret': True}, path=['password']
        schema_item='string', path=['password', 'type']
        schema_item=True, path=['password', 'airbyte_secret']
        """
```